test/backend-cpp/TestHelper.hpp

```diff
@@ -0,0 +1,86 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//====----------------------- TestHelper.hpp ------------------------------===//
+//
+// Copyright 2022 The IBM Research Authors.
+//
+// =============================================================================
+//
+// This file defines a set of helper functions for creating, compiling and
+// running an end-to-end test.
+//
+//====---------------------------------------------------------------------===//
+
+#ifndef ONNX_MLIR_TEST_HELPER_H
+#define ONNX_MLIR_TEST_HELPER_H
+
+#include "src/Compiler/CompilerUtils.hpp"
+#include "src/Dialect/ONNX/ONNXOps.hpp"
+#include "src/Runtime/ExecutionSession.hpp"
+#include "src/Runtime/OMTensorHelper.h"
+
+#include <memory>
+#include <string>
+#include <vector>
+
+using namespace std;
+using namespace mlir;
+
+namespace BackendCppTests {
+
+// Helper class containing useful function for creating, compiling and running a
+// test.
+class TestHelper {
```

Comment:
I like the idea of this, maybe the interface could be simplified a bit.  This is what I "implemented" informally in the model library, but without classes

* Build and compile
* Run
* Compare.

and since the build and compile contains all the loc, context, module. builder, there was no need to expose them.

There is a certain elegance in having these var in the class, as then you can have support operations. 
I can see the following process:
1. Build the graph (using helper function). One issue is that for some operations, we need to fill in weights (which are randomly selected numbers). These "constants" needs then to be passed to the computations for the expected values. But these constants can be added to the class, so that is a neat way to do it. So you have here an "abstract builder", and each subclass implements the actual graph. This builder functions would have many parameters specialized for the actual operations, e.g. matmul needs I,J,K, convolution needs N C H W ...
2. compile, that can be done in the parent class
3. compute expected: that would be specialized by each of the subclass, and result in the vector of expected value
4. run: again that can be done by the superclass
5. compare results: you can have implementations of comparison for each type of results, currently float or string

So I like this idea, I would rename this "ModelBuilder" and combine the ones from the numerical and this one into a single library of models we can build, and then have very lightweight process to do the rapid check (lots of tests) vs backend-cpp which is a few tests.

Reply:
> and since the build and compile contains all the loc, context, module. builder, there was no need to expose them.

Several tests would need the same functionality to create a test case. So I have provided member functions to facilitate creating the skeleton for testing an operator. Users of `TestHelper` would then just need to provide the functionality to create the operator that needs to be tested, and rely on the TestHelper class to deal with the skeleton (create a module, create a function in the module, push a return operator in the function, etc..)

Reply:
> compare results: you can have implementations of comparison for each type of results, currently float or string

I can address this point by allowing the user of `ModelBuilder` to provide a function object where the verification code is "encoded". That way the consumer can customize the verification code appropriately for the particular scenario it wants to test.